Improve performance of `ListResources`

rosstimothy opened the pull request:
Reduces latency of `proto.AuthService/ListResources` in a variety of ways:

- 6f9ac02: Moves RBAC logging from DEBUG to TRACE 
- c71ef7b: Stores compiled `*regexp.Regexp` in a LRU cache so that they can be reused during RBAC
- ec0860c: Adds a `GetLabel(key string) (value string, ok bool)` to `types.ResourcesWithLabels` to prevent copying the entire label set when we just need to look up keys
- 78d0bef: Prevents loading an extra page from the cache to determine the next key in `auth.ServerWithRoles.ListResources`
- 81b08ed: Modifies `services.UnmarshalServer` to unmarshal directly into a `types.ServerV2` instead of first into a `types.ResourceHeader` to check that the version is `types.V2`


Comparison of `BenchmarkListNodes` from https://github.com/gravitational/teleport/pull/23534/commits/b1715a5d3bb804428efbf9b16def5778c4e4ee5f to https://github.com/gravitational/teleport/pull/23534/commits/ec0860c72047666b533a174a6ea2e1931a87be8f:
```
benchstat b1715a5.txt ec0860c.txt
goos: darwin
goarch: amd64
pkg: github.com/gravitational/teleport/lib/auth
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
             │ b1715a5.txt │             ec0860c.txt             │
             │   sec/op    │   sec/op     vs base                │
ListNodes-16   19.307 ± 5%   1.033 ± 14%  -94.65% (p=0.000 n=10)

             │  b1715a5.txt   │             ec0860c.txt              │
             │      B/op      │     B/op      vs base                │
ListNodes-16   11154.5Mi ± 0%   493.0Mi ± 0%  -95.58% (p=0.000 n=10)

             │  b1715a5.txt  │             ec0860c.txt             │
             │   allocs/op   │  allocs/op   vs base                │
ListNodes-16   112.067M ± 0%   8.341M ± 0%  -92.56% (p=0.000 n=10)
```

rosstimothy committed:
Add benchmark for ListNodes

rosstimothy committed:
Intern compiled regular expressions

Profiles of the benchmark test revealed that the `regexp.Compile`
done within `utils.matchString` was the most cpu and memory intensive
portion of the tests. By leveraging a `lru.Cache` to intern the
compiled regular expressions we get quite a performance improvement.

rosstimothy committed:
Only fetch a single page of resources

Increases the request limit prior to loading the resources from
the cache so that we load enough items in a single page to determine
the start key of the next page.

rosstimothy committed:
Remove version checking from `services.UnmarshalServer`

Unmarshal directly to a `types.ServerV2` instead of first creating
a `types.ResourceHeader` to inspect the version. There is only a
single version for `types.ServerV2` making the check unnecessary.

rosstimothy committed:
Add `GetLabel` to `types.ResourceWithLables`

`GetAllLabels` can be overkill if one simply needs to look up the
value for a particular label. It creates a new `map[string]string`
and copies all of a resources existing labels. RBAC decisions
driven by labels incurred the penalty of the copy each time access
was checked. The impact of the copy is much more noticeable when
a resource has several labels or really long strings in the key
or value.

By leveraging `GetLabel` RBAC can avoid copying the labels altogether
and simply lookup each label key when required.

fspmarshall reviewed (commented):
nit: the old `CombineLabels` strategy would perform the combination s.t. command labels took precedence over static labels (i.e. if a command label and a static label exist for the same key, only the command label would be observed).  As implemented, these `GetLabel` methods are giving precedence to static labels.  I actually prefer this strategy, but it would technically be a breaking change in RBAC behavior, so probably best to change it.

rosstimothy committed:
Move RBAC logging to trace level

BenchmarkListNodes is twice as slow when RBAC logging is enabled.
By switching RBAC logging from debug to trace we can eliminate
the performance hit while still providing a way for users to opt
in to the behavior if they need to debug RBAC.

zmb3 reviewed (approved):
nice work!

strideynet commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -1144,6 +1148,107 @@ func TestSessionRecordingConfigRBAC(t *testing.T) {
+		values = append(values, uuid.New().String())
+	}
+
+	values[0] = "evil"
```
Nit: I know this is just a benchmark but think we could probably use something a bit clearer than "evil" here. Perhaps "hidden" or "inaccessible" ?

strideynet reviewed (approved):
Great work, awesome to see benchmarks being used to ensure improvements are worthwhile.
